packages/components/src/presentation-components/properties/UniquePropertyValuesTargetSelector.tsx

```diff
@@ -0,0 +1,154 @@
+  if (property && property.valueFormat === PropertyValueFormat.Primitive && typeof property.value === "string" && property.displayValue) {
+    const { displayValue, value } = property;
+    const parsedDisplayValues = JSON.parse(displayValue);
+    const parsedGroupedRawValues = JSON.parse(value);
+    const uniqueValues: DisplayValueGroup[] = [];
+    for (let i = 0; i < parsedDisplayValues.length; i++) {
+      uniqueValues.push({ displayValue: parsedDisplayValues[i], groupedRawValues: parsedGroupedRawValues[i] });
+    }
+    return uniqueValues;
+  }
+  return undefined;
+}
+
+interface UseUniquePropertyValuesRulesetProps {
+  property: PropertyDescription;
+  fieldDescriptor?: FieldDescriptor;
+}
+
+function useUniquePropertyValuesRuleset({ property, fieldDescriptor }: UseUniquePropertyValuesRulesetProps) {
+  const [ruleset, setRuleset] = useState<Ruleset>();
+  const [schemaName, className] = fieldDescriptor && FieldDescriptor.isProperties(fieldDescriptor) ? fieldDescriptor.properties[0].class.split(":") : ["", ""];
+  useEffect(() => {
+    setRuleset({
+      id: "unique-property-values",
+      rules: [
+        {
+          ruleType: RuleTypes.Content,
+          specifications: [
+            {
+              specType: ContentSpecificationTypes.ContentInstancesOfSpecificClasses,
+              classes: { schemaName, classNames: [className], arePolymorphic: true },
+            },
+          ],
+        },
+      ],
+    });
+  }, [property, schemaName, className]);
+
+  return ruleset;
+}
+
+interface UseUniquePropertyValuesLoaderProps {
+  imodel: IModelConnection;
+  ruleset?: Ruleset;
+  fieldDescriptor?: FieldDescriptor;
+}
+
+function useUniquePropertyValuesLoader({ imodel, ruleset, fieldDescriptor }: UseUniquePropertyValuesLoaderProps) {
+  const loadTargets = useCallback(
+    async (_: string, loadedOptionsCount: number) => {
+      if (!ruleset || !fieldDescriptor) {
+        return { options: [], hasMore: false };
+      }
+
+      const content = await Presentation.presentation.getPagedDistinctValues({
+        imodel,
+        descriptor: {},
+        fieldDescriptor,
+        rulesetOrId: ruleset,
+        paging: { start: loadedOptionsCount, size: PRESENTATION_TARGETS_BATCH_SIZE },
+        keys: new KeySet(),
+      });
+      return {
+        options: content.items.filter(
+          (item) =>
+            item.groupedRawValues.filter((value) => value !== undefined && value !== null) && item.displayValue !== undefined && item.displayValue !== "",
+        ),
+        hasMore: content.items.length === PRESENTATION_TARGETS_BATCH_SIZE,
+      };
+    },
+    [imodel, ruleset, fieldDescriptor],
+  );
+
+  return loadTargets;
+}
+
+function stringifyGroupRawValuesTo(values: MultiValue<DisplayValueGroup>) {
+  const singleArray: Value[] = [];
+  values.forEach((item) => singleArray.push(item.groupedRawValues));
+  return JSON.stringify(singleArray);
```

Comment:
I see there are multiple places where values are serialized deserialized. Looks like it would be worth introducing some functions that would serialize/deserialize `MultiValue` and could be reused in all different places

Reply:
fixed